### PR TITLE
cmd/protoc-gen-go-grpc: rework service registration

### PR DESCRIPTION
**Pull request by dfawley**
Sadly there is a large amount of churn in generated code in this PR, due to changing over to the legacy codegen for everything besides examples to avoid breaking changes.  If new methods are added to any services exposed by them, we will switch to the new version of the tool since we would be forced to break compatibility no matter what.

---
This change is [Reviewable](https://reviewable.io/reviews/grpc/grpc-go/3828)

### CODE REVIEW

File: cmd/protoc-gen-go-grpc/README.md [hunk 1]
```diff
@@ -4,18 +4,115 @@ This tool generates Go language bindings of `service`s in protobuf definition
...
+services.  In that version, it was only possible to register a service
+implementation that was a complete implementation of the service.  It was also
+possible to embed an `Unimplemented` implementation of the service, which was
+also generated and returned an UNIMPLEMENTED status for all methods.  To make
+it easier to migrate from the previous version, two new symbols are generated:
+`New<Service>Service` and `Unstable<Service>Service`.
+
+`New<Service>Service` allows for easy wrapping of a service implementation into
+an instance of the new generated `Service` struct type.  *This has drawbacks,
+however: `New<Service>Service` accepts any parameter, and does not panic or
```
Comment (easwars) [inline cmd/protoc-gen-go-grpc/README.md:75]: If methods are missing or misspelt, does it eventually return UNIMPLEMENTED at runtime? Maybe document what the expected behavior is in that case.

Comment (easwars) [reply cmd/protoc-gen-go-grpc/README.md:75]: I just the generated code, which does say that missing/misspelt methods will return UNIMPLEMENTED. Maybe it makes sense to document that here as well.

Comment (dfawley) [reply cmd/protoc-gen-go-grpc/README.md:75]: Done

**Comment by menghanl** [comment]
Per offline discussion, try keeping both old and new symbols (RegisterEchoServer and RegisterEchoService), so there is a migrating path.

**Pull request merged**